### ISSUE TITLE
[RFC] Should PAL stop proxying Carbon components that we do not extend?

### ISSUE THREAD

**Issue by davidicus** [carbon-design-system/carbon-addons-iot-react#2445]
### What package is this for?

- [x] React

### Summary

With our upcoming v3 release we have the opportunity to improve the project and the way our ecosystem works. One item up for change is the proxying of all Carbon components that we DO NOT extend. Up to this point we have been a single dependency that projects need to manage and the consumption of Carbon was done through the `carbon-addons-iot-react` package. A proposed change is to stop exporting Carbon components that are not extended by the PAL. 

### Original Reasoning to proxy
Carbon ships multiple packages that can be used by teams to help build out their UIs while maintaining the IBM styles guide. With people just getting started knowing which packages to use and managing those dependencies was a difficulty we wanted to alleviate. Additionally there was some confusion as where to open up bugs and if those bugs stemmed from Carbon or originated in the PAL extensions.  The PAL team decided on being the sole dependency where teams could install and manage one single package. This effectively made the PAL the official way the applications in the AI Applications org could consume Carbon and modernize their UIs. This also limited the places where app devs needed to go to find documentation. 

### The problems with this approach
As both the Carbon and PAL libraries grow and as more applications start to use the PAL package we have discovered a few issues with this approach. 

- Added complexity for builds:
Dealing with pulling in dependencies and exporting them has lead to inconsistencies and some times circular dependency issues. This just makes it harder for the PAL team to properly maintain the code we do write and clogs up the console with unnecessary noise. 
- Our package routinely has breaking changes that are out of our control:
Carbon has a different philosophy around Semantic versioning. This leads to breaking changes for our consumers inside a minor PAL release. I believe this negatively impacts our consumer's opinion of the project and its stability. Although, we will still have this with items we extend the amount of breakages that consumers perceive to be PAL specific will go down. I think this will also help to identify where the issue originates.
- Clarity on what value the PAL provides beyond base Carbon:
It may be hard to discern what features are being provided from Carbon and what comes from PAL since users only interact with the PAL project. When trying to run reports, digest the extended components or figure out where bugs originate are all made more difficult by us proxying the Carbon components.
- Having to support docs for ALL their components:
This has been recently made more difficult with Carbon no longer exporting their stories or referencing other resources that are in their own project repo. We now must do a manual update to our stories by copying over and tweaking the Carbon version. This is error prone since it is manual and can easily get out of date if stories are update on the Carbon side but not updated in the PAL storybook
- We have to manually update our imports/exports when new components are added:
Carbon library is constantly growing. When new components are added they can easily be missed and not exported in the PAL library. There are some test we could put in place that would test all the exports from Carbon and the amount of components we export but it is error prone and still requires a manual update.

### Proposed step to have application teams consume Carbon directly 
It may be beneficial to ask application teams to consume Carbon directly. While many of the issues above are really issues for PAL maintainers, the complexity it adds can still lead to negative results for consumers of the PAL package. There are some negatives with this approach and this RFC is a request to debate the pros and cons to both solutions. 

**Some pros:**
- Less complexity for the project:
PAL maintainer can focus on making our code more robust and reliable without having to deal with a large number of components that we do not extend.
- Better visibility:
This means better visibility of what the PAL provides, where bugs originate from, what the real dependency list is, and better documentation for Carbon components that are not extended. 
- New Components are automatically available. 
There is no manual step to import and export new components. As soon as Carbon publishes them and PAL upgrades those components are available for consumers. 
- Documentation stays up to date:
Without having to manually copy stories from Carbon the docs that people use will be up to date. 
- Contributors would feel more comfortable working in the project:
With less clutter in the project new comers would have an easier time getting acclimated and take less time to get a "feel" for the project

**Some Cons:**
- Everyone must deal with the breaking changes from Carbon:
If breaking changes happen there are times when the PAL can provide the fixes ourselves so each downstream consumer does not have to deal with them. This will not stop all  breaking changes from making its way to product teams but can help reduce the number. 
- Multiple dependencies to manage: 
While I am proposing that we create CLI tools that will help with installing and maintaining compatible  deps there may be those who are unaware or just not comfortable using the CLI and will have to manually update these dependencies.
- Import paths will have to be updated:
Similarly there are code mod scripts that we can provide that will help with this task but these scripts are not fool proof and edge cases may cause some issues.

### Plans to implement
If we were to indeed move forward with this plan there are multiple things that we would have to provide or do to make this change.
- Code mods to help changing of application imports
- CLI script for installing and maintaining dependencies for Carbon/PAL ecosystem
- Update our docs to point to the Carbon version
- Update our codebase to remove the exports.

### Summary
I think there are a lot of positives to removing this complexity to the project but also understand that this may shift complexity to product developers. The thing we should decide is what would allow all of us to ship better, more reliable UIs.

**Comment by stuckless** [comment]
I'd be ok with this.  I think the initial goal of wrapping carbon was to bring stability of Carbon to downstream consumers, but that never really happened, and the PAL team generally takes the brunt of that.  This doesn't really solve our problem, but, it likely won't make it much worse either, but it will simplify PAL.

**Comment by davidicus** [comment]
> I'd be ok with this. I think the initial goal of wrapping carbon was to bring stability of Carbon to downstream consumers, but that never really happened, and the PAL team generally takes the brunt of that. This doesn't really solve our problem, but, it likely won't make it much worse either, but it will simplify PAL.

Yes it definitely does not fix the breakages. I am thinking the simplification of the PAL code base may provide more positive benefits than the few fixes that we are able to intercept and resolve. My biggest concern are missing out on new components shipped from Carbon and documentation getting out of date. There may be things we can do with the documentation to improve that but not sure the extent of how complex that could get.

**Comment by JordanWSmith15** [comment]
@davidicus  Given the lack of response from the PAL slack question, I think we are free to do this if we want/if it bubbles up to the top of the list

**Comment by daphoenix** [comment]
What does your target user base need?

What's the benefits of not proxying the non-extended, default carbon components?
- Simpler packaging for the PAL team

What's the disadvantages to of not proxying the non-extended, default carbon components?
- Another dependency for the any developer using these components manage. For product teams that's another opensource software process.
- Debugging would be harder

Whilst it is complex, it's easier for a PAL team to manage and communicate the scope of this issue to the Carbon team, than individual development teams.

I'm currently using this add-on with my product team.
